### PR TITLE
fix memory leak in metrics reporter. make apikey thread-safe

### PR DESCRIPTION
**Pull request by njayp**
Signed-off-by: njayp <nickjaypowell@gmail.com>
Fixed the memory leak. Note that `comm != nil` is not thread-safe, this will be addressed in a future PR

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -816,75 +810,88 @@ func (a *Agent) ProcessDiagnostics(ctx context.Context, diagnostics *diagnostics
...
+			return err
+		case <-time.After(defaultMinReportPeriod):
+			a.ReportMetrics(ctx)
 		}
+	}
+}
 
-		instanceID := p.Addr.String()
+func (a *Agent) ReportMetrics(ctx context.Context) {
+	// TODO make thread-safe
```
Comment (raphaelreyna) [inline pkg/agent/agent.go:829]: This should be made thread-safe

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -816,75 +810,88 @@ func (a *Agent) ProcessDiagnostics(ctx context.Context, diagnostics *diagnostics
...
+	}
+}
+
+// MetricsRelayHandler is invoked as a callback when the agent receive metrics from Envoy (sink).
+// It stores metrics in a.aggregatedMetrics
+func (a *Agent) MetricsRelayHandler(
+	ctx context.Context,
+	in *envoyMetrics.StreamMetricsMessage,
+) {
+	// TODO make thread-safe
```
Comment (raphaelreyna) [inline pkg/agent/agent.go:865]: this too

**Review by raphaelreyna — CHANGES_REQUESTED** [review]
looks like there are still some TODOs

File: cmd/main.go [hunk 3]
```diff
@@ -28,8 +28,8 @@ import (
 // internal k8s service
 const (
 	AdminDiagnosticsPort     = 8877
-	DefaultSnapshotURLFmt    = "http://ambassador-admin:%d/snapshot-external"
-	DefaultDiagnosticsURLFmt = "http://ambassador-admin:%d/ambassador/v0/diag/?json=true"
+	DefaultSnapshotURLFmt    = "http://edge-stack-admin:%d/snapshot-external"
```
Comment (LanceEa) [inline cmd/main.go:31]: This change is fine for Helm but breaks the default standalone yaml shipped with AES and Emissary-ingress (maybe). This needs to be coordinated with changes or probably a better long term approach would be to move away from Edgissary maintaining its own stuff and depend on this repository to own it.

Comment (LanceEa) [reply cmd/main.go:31]: So since it is not necessary to fix the memory-leak can this be extracted into a separate PR and tested against AES helm and yaml?

Comment (njayp) [reply cmd/main.go:31]: yep

**Review by raphaelreyna — APPROVED** [review]

**Pull request merged**